### ISSUE TITLE
How to build the relationship between blocks？

### ISSUE THREAD

**Issue by TediousBoredom** [kuleshov-group/bd3lms#15]
class DiTBlockCausal(nn.Module) And class DDiTBlock(nn.Module) in the code use the time_condition_embed  and block_diff_mask generated by and gen_mask function as mask respectively, but as you mentioned in the paper, you apply the NELBO of diffusion to each term in likelihood of blocks, for a sequence x as input, what's the details of modeling the x into several blcoks. 

The  forward process of class DIT(nn.Module, huggingface_hub.PyTorchModelHubMixin) process the context sequence of length L .How to build the relationship between blocks？ does the DIT class mean process the x into blocks and just use the function cross_attention to model the casual block generation given the blcoks that are <b

**Comment by mariannearr** [comment]
We model multiple blocks simultaneously as follows:
- As described in Alg 1 from our paper, we sample a different noise level for each block. This is done by `_sample_t` in `diffusion.py`, where we output a single tensor `t` of shape `[batch_size, seq_len]` where the noise level is the same for tokens in the same block.
- The loss coefficient `loss_scale` is just `1/t`, so the loss scaling will be the same for tokens belonging to the same block. This is handled in the `LogLinear` noise instance in `noise.py`
- Our objective applies the diffusion NELBO within each block, but we can pass the entire sequence into the transformer and simultaneously (using a custom attention mask)
- The relationship between blocks is enforced through the attention mask (pictured below)

![Image](https://github.com/user-attachments/assets/03a0b0ed-3db4-4638-949c-bb38e1245948)

. This is created using `block_diff_mask` in `models/dit.py`. You can also find further details in Suppl. B7 from our paper

Hope this helps.
